hugolib/site_sections.go

```diff
@@ -0,0 +1,278 @@
+// To get a section's subsections, see Page's Sections method.
+func (p *Page) Parent() *Page {
+	if p.Kind == KindPage {
+		return p.currentSection
+	}
+	return p.parent
+}
+
+// current returns the page's current section.
+// Note that this will return nil for pages that is not regular, home or section pages.
+// Note that for paginated sections and home pages, this will return the original page pointer.
+func (p *Page) current() *Page {
+	v := p
+	if v.origOnCopy != nil {
+		v = v.origOnCopy
+	}
+	if v.IsHome() || v.IsSection() {
+		return v
+	}
+
+	return v.currentSection
+}
+
+// InSection returns whether the given page is in the current section.
+// Note that this will always return false for pages that are
+// not either regular, home or section pages.
+func (p *Page) InSection(other interface{}) (bool, error) {
+	if p == nil || other == nil {
+		return false, nil
+	}
+
+	if po, ok := other.(*PageOutput); ok {
+		other = po.Page
+
+	}
+
+	pp, ok := other.(*Page)
+	if !ok {
+		return false, fmt.Errorf("%T not supported in IsActive", other)
+	}
+
+	if pp == nil {
+		return false, nil
+	}
+
+	return pp.current() == p.current(), nil
+}
+
+// Sections returns this section's subsections, if any.
+// Note that for non-sections, this method will always return an empty list.
+func (p *Page) Sections() Pages {
+	return p.subSections
+}
+
+func (s *Site) assembleSections() Pages {
+	var newPages Pages
+
+	if !s.isEnabled(KindSection) {
+		return newPages
+	}
+
+	// Maps section kind pages to their path, i.e. "my/section"
+	sectionPages := make(map[string]*Page)
+
+	// The sections with content files will already have been created.
+	for _, sect := range s.findPagesByKind(KindSection) {
+		sectionPages[path.Join(sect.sections...)] = sect
+	}
+
+	const (
+		sectKey     = "__hs"
+		sectSectKey = "_a" + sectKey
+		sectPageKey = "_b" + sectKey
+	)
+
+	var (
+		home       *Page
+		inPages    = radix.New().Txn()
+		inSections = radix.New().Txn()
+	)
+
+	for i, p := range s.Pages {
+		if p.Kind != KindPage {
+			if p.Kind == KindHome {
+				home = p
+			}
+			continue
+		}
+
+		sectionPath := path.Join(p.sections...)
+		sect, found := sectionPages[sectionPath]
+
+		if !found && len(p.sections) == 1 {
+			// We only create content-file-less sections for the root sections.
+			sect = s.newSectionPage(p.sections[0])
+			sectionPages[sectionPath] = sect
+			newPages = append(newPages, sect)
+		} else if !found && len(p.sections) > 0 {
+			// If it isn't a sub-section then it is a section from the
+			// first path element.
+			sectionPath = p.sections[0]
+			sect, found = sectionPages[sectionPath]
+			if !found {
+				sect = s.newSectionPage(sectionPath)
+				sectionPages[sectionPath] = sect
+				newPages = append(newPages, sect)
+			}
+		}
+
+		pagePath := path.Join(sectionPath, sectPageKey, strconv.Itoa(i))
+		inPages.Insert([]byte(pagePath), p)
+	}
+
+	s.Pages = append(s.Pages, newPages...)
+
+	for k, sect := range sectionPages {
```

Comment:
You range over `sectionPages` 4 times in this func.  It looks like you could cut that down to 2.